lib/babe/build.go

```diff
@@ -189,11 +189,30 @@ func (b *BlockBuilder) buildBlockSeal(header *types.Header) (*types.SealDigest,
 func (b *BlockBuilder) buildBlockExtrinsics(slot Slot, rt runtime.Instance) []*transaction.ValidTransaction {
 	var included []*transaction.ValidTransaction
 
-	for !hasSlotEnded(slot) {
+	slotEnd := slot.start.Add(slot.duration * 2 / 3) // reserve last 1/3 of slot for block finalisation
+	timeout := time.Until(slotEnd)
+	slotTimer := time.NewTimer(timeout)
+
+	for {
+		select {
+		case <-slotTimer.C:
+			return included
+		default:
+		}
+
 		txn := b.transactionState.Pop()
-		// Transaction queue is empty.
-		if txn == nil {
-			continue
+		retry := txn == nil // Transaction queue is empty.
```

Comment:
I think this code would be more straight forward if we had some sort of `TransactionState.PopChannel() (tx *transaction.ValidTransaction, cancel func() (err error)` method.  That way you can move the retry logic within the `TransactionState`. The calling code would look a lot simpler:
```
	for {
                popChan, cancel := b.transactionState.PopChannel()
		select {
		case <-slotTimer.C
		        // ensure that popChan is closed
		        var wg sync.WaitGroup
		        wg.Add(1)
		        go func(){ 
		            <-popChan
                             wg.Done()
		         }()
		        cancel()
		        wg.Wait()
			return included
		case txn := <-popChan:
			// handle txn
			extrinsic := txn.Extrinsic
			...
		}
        }
```

Reply:
Good point, we could just return a channel from the Pop method.
Although we should probably just have one Pop method (and not one returning nil, one returning a channel) to avoid any data race / reduce complexity.

a few nits on that:

- the method is `TransactionState.PopChannel() (tx <-chan *transaction.ValidTransaction, cancel func())` I think
- Use `done := make(chan struct{})`, not empty interface which does occupy a few bytes vs struct{}

```go
		done := make(chan interface{})
		go func(){ 
			<-popChan
			close(done)
		}()
		cancel()
		<-done
```

can just be

```go
		cancel()
		<-popChan
```

Reply:
I've refactored this to create function `TransactionState.PopChannel(*time.Timer) (tx chan *transaction.ValidTransaction)` which take a time that will cancel the channel when the timer expires.  This seems like a good solution that eliminates the need for a cancel function.